### PR TITLE
Register factories custom elements to avoid duplicate creation calls

### LINKED ISSUE #80: Race condition with widget creation by type
When `getWidget` is called for an `id` that is dynamically created using the `.type` from the associated to the state entry. If `getWidget` is called again (with the same id) before the widget has actually been factoried and registered then `app` will try and create the widget again. 

This causes a failure when attempting to register the (duplicate) widget as an entry for that `id` already exists in the registry.

**Version and Environment:**

Dojo2 version: latest

**Code**

```ts
import createApp from 'dojo-app/createApp';
import createMemoryStore from 'dojo-stores/createMemoryStore';

import createWidget from 'dojo-widgets/createWidget';
import createPanel from 'dojo-widgets/createPanel';

const widgetStore = createMemoryStore<any>({
	data: [
		{
			id: 'container'
		}
	]
});

const app = createApp({ defaultWidgetStore: widgetStore });

app.loadDefinition({
	widgets: [
		{
			id: 'container',
			factory: createPanel
		}
	],
	customElements: [
		{
			name: 'dynamically-created-child',
			factory: createWidget
		}
	]
});

app.realize(document.body)
	.then(() => {
		const children: string[] = [];
		for (let i = 0; i < 10; i++) {
			const id = `${i}`;
			children.push(id);
			widgetStore
				.put({id, type: 'dynamically-created-child', label: 'This is label number ' + i})
				.patch({id: 'container', children});

		}
	});

```

**Expected behavior:**

Render 10 children items:

```
This is label number 0
This is label number 1
This is label number 2
This is label number 3
This is label number 4
This is label number 5
This is label number 6
This is label number 7
This is label number 8
This is label number 9
```

**Actual behavior:** 

No children render.

### PR DESCRIPTION
**Pull request by agubler**
**Type:** bug

**Description:** 

Register factories for dynamically created custom factories so that if the same widget is requested during creation a promise is returned instead of attempting to create the widget twice.

**Related Issue:** #80 

fixes #80

Please review this checklist before submitting your PR:

* [x] There is a related issue
* [x] All contributors have signed a CLA
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] The code passes the CI tests
* [x] Unit or Functional tests are included in the PR
* [x] The PR increases or maintains the overall unit test coverage percentage
* [x] The code is ready to be merged

### CODE REVIEW

**Comment by agubler** [comment]
@novemberborn I think this should do the trick, I'll add some tests to prove the scenario... Please let me know if you were thinking of something different.

**Comment by novemberborn** [comment]
> @novemberborn I think this should do the trick, I'll add some tests to prove the scenario... Please let me know if you were thinking of something different.

No this is great, simpler than what I was thinking of!

**Comment by agubler** [comment]
@novemberborn I've been thinking, do you think we could refactor this a little? I was thinking that instead of being reactive and waiting for `getWidget` to be called before determining whether we can create an instance of the widget. Do you think that `app` could observe the store(s) registered (so enabling support for non the default widgetStore) and if a `type` exists on the state record then look up the custom element and register the factory? That will simplify the `getWidget` method and also enable us to revert back to a synchronous `hasWidget` check? It would also eliminate the (admittedly small) possibility that even with this guard users could get a failure as two of the same `id`/`widget` are attempted to be created and registered?

**Comment by novemberborn** [comment]
> Do you think that app could observe the store(s) registered (so enabling support for non the default widgetStore) and if a type exists on the state record then look up the custom element and register the factory?

Continuously registering factories sounds more complicated than doing it lazily. Especially when it comes to `type` mutations, in which case, what if a widget has already been instantiated? Or if `type` is used for another purpose?

> It would also eliminate the (admittedly small) possibility that even with this guard users could get a failure as two of the same id/widget are attempted to be created and registered?

I don't see how you would get two separate instances. You cannot register a factory that's already there.

**Comment by agubler** [comment]
>Continuously registering factories sounds more complicated than doing it lazily. Especially when it comes to type mutations, in which case, what if a widget has already been instantiated? Or if type is used for another purpose?

It would only attempt to register a factory when an `widgetStore` item is created and only if it has the "special" `type` field that has a value that matches the name of a `customElement` that has been registered.

If the `type` changes then that could be complicated to handle... I would need to think about that some more.

> I don't see how you would get two separate instances. You cannot register a factory that's already there.

The instances do get created but you will get the failure when the second widget instance is registered (as is happening now).

**Comment by novemberborn** [comment]
> It would only attempt to register a factory when an widgetStore item is created and only if it has the "special" type field that has a value that matches the name of a customElement that has been registered.

What if a custom element is registered later?

> The instances do get created but you will get the failure when the second widget instance is registered (as is happening now).

To avoid the failure you could synchronously check whether a factory is registered for the `id` and only register when that's not the case. That check is done elsewhere but I suppose concurrent calls can race to reach `createCustomWidget`. Will be hard to test for that though!

**Comment by agubler** [comment]
>What if a custom element is registered later? 

I think it could only handle store for already registered custom elements (just as calling `getWidget` with a type before the custom element is registered wouldn't create the widget), but from that point onwards any store items updated/added would register the new custom element as their factory.

>To avoid the failure you could synchronously check whether a factory is registered for the id and only register when that's not the case. That check is done elsewhere but I suppose concurrent calls can race to reach createCustomWidget. Will be hard to test for that though!

It would be fine but we need to look up the state from the store to determine the factory to use and it is that async'ness that could cause the check in `getWidget` to not return the factory for the second call and go through the `createCustomElement` path again. Actually we could do another check just beforehand (after fetching the state record) that would actually remove that small edge case (that's what you may have been suggesting!)

**Comment by agubler** [comment]
@novemberborn the other conversation aside, I've updated the PR to include a test with concurrent calls.

File: src/createApp.ts [hunk 1]
```diff
@@ -569,17 +569,28 @@ function addIdentifier(app: App, id: Identifier) {
...
 	const { registryProvider, defaultWidgetStore } = app;
+	let factoryHandle: Handle;
 	// istanbul ignore if
 	if (!defaultWidgetStore) {
 		throw new Error('A default widget store must be configured in order to create custom widgets');
 	}
 
 	return defaultWidgetStore.get(id).then((state: any) => {
-		const options: WidgetFactoryOptions = { id, stateFrom: defaultWidgetStore, registryProvider, state };
 		const customFactory = customElementFactories.get(app).get(state.type);
```
Comment (novemberborn) [inline src/createApp.ts]: Just realizing now it'd be nice if this checks whether `state.type` even has a (string) value.

File: src/createApp.ts [hunk 2]
```diff
@@ -569,17 +569,28 @@ function addIdentifier(app: App, id: Identifier) {
...
-		return customFactory(options);
+		const factories = widgetFactories.get(app);
+		let factory: RegisteredFactory<WidgetLike>;
+		try {
+			// add final check before attempting to register the factory
```
Comment (novemberborn) [inline src/createApp.ts]: If you check `factories.has(id)` and `widgetInstances.get(app).has(id)` instead you won't need the `try/catch`.

Comment (novemberborn) [inline src/createApp.ts]: Then make the registration conditional on that. If the widget has already been registered don't register the new factory.

File: src/createApp.ts [hunk 3]
```diff
@@ -569,17 +569,28 @@ function addIdentifier(app: App, id: Identifier) {
...
+			factory = factories.get(id);
+		}
+		catch (err) {
+			factoryHandle = app.registerWidgetFactory(id, customFactory);
+			factory = factories.get(id);
+		}
+
+		return factory();
```
Comment (novemberborn) [inline src/createApp.ts]: And return `app.getWidget(id)` here. Works whether you need to use an existing factory or the freshly registered custom factory.

File: src/createApp.ts (hunk 1 above)
Comment (agubler) [inline src/createApp.ts]: Will follow up in a separate PR
